site/_posts/2018-12-12-reinventing-the-proxy.md

```diff
@@ -0,0 +1,124 @@
+
+```ruby
+class Proxy
+  def push(ws_sock, console_sock)
+    Thread.new do
+      loop do
+        data = ws_sock.read(1024)
+        console_sock.write(translate_ws(ws_data))
+      end
+    end
+
+    Thread.new do
+      loop do
+        data = sock.read(1024)
+        ws_sock.write(translate_console(data))
+      end
+    end
+  end
+end
+
+```
+The problems appear when having more clients, e.g. with the SPICE protocol that require more than a single connection to operate. By using non-blocking read and write it is enough to have a single thread that implements the [reactor pattern](https://en.wikipedia.org/wiki/Reactor_pattern). This thread should run two repeating operations in a loop: testing the readiness of sockets and transmitting data between the ready ones in a non-blocking way:
+
+```ruby
+class Proxy
+  def initialize
+    @pairing = {}
+
+    Thread.new do
+      loop do
+        reads, writes, _ = IO.select(@sockets, @sockets, [], 0.1)
+        # Iterate through the reads
+        read.each do |reads|
+          # Retrieve the other end and the translator method from the pairing hash
+          write, translate = @pairing[read]
+
+          # Skip the transmission if the other end is not writable
+          next unless writes.include?(writes)
+
+          # Do the transmission in a non-blocking way
+          data = read.read_nonblock(1024)
+          write.write_nonblock(translate.call(data))
+        end
+      end
+    end
+  end
+
+  def push(ws_sock, console_sock)
+    @sockets.push(ws_sock, console_sock)
+
+    # Pair the sockets and the translation methods
+    @pairing[ws_sock] = [console_sock, method(:translate_ws)]
+    @pairing[console_sock] = [ws_sock, method(:translate_console)]
+  end
+end
+```
+
+This implementation is really close to what we used to have in the last two years, of course with additional mutexes and error handling. As you can see, the `IO.select` was utilized to determine if a socket is ready for reading and/or writing. However, this implementation can cause a [spinlock](https://en.wikipedia.org/wiki/Spinlock) that is encoded into the usage of `IO.select` inside the `loop`. The last argument of the `IO.select` method is a timeout which states how many seconds should the should thread wait if no socket is ready. This should cause our thread to sleep when there is nothing to transmit and so the computer can do other useful stuff in the meantime. The problem with this is that the `IO.select` will return from its sleep whenever a single socket is ready for transmission. In other words, `IO.select` cannot be forced to wait longer for a pair of sockets to become ready for reading and writing at least in one direction.
```

Comment:
"how many seconds should the should thread wait" -> are there too many `should`s?